docs/specs/getAgent.md

```diff
@@ -0,0 +1,146 @@
+
+The WCP algorithm (coordinated between `getAgent()` and DAs) has four steps. Each step may contain sub-steps.
+
+1. Connect to DA
+2. Validate app identity
+3. Persist DesktopAgentDetails to SessionStorage
+4. Resolve promise with DesktopAgent interface
+
+### Connect to DA (Step 1)
+
+1) Check for a possible navigation or refresh event
+
+    Check the SessionStorage key `FDC3-Desktop-Agent-Details` for a `DesktopAgentDetails` record. If it exists **and the appId matches**, then a navigation or refresh event has occurred and the stored data MUST be used to reestablish a connection to the DA. This ensures that the window maintains a consistent `instanceId` between navigation events within an app. If it doesn't exist, or the appId does not match, then proceed to step (2).
+
+    See [DesktopAgentDetails](./getAgent.d.ts).
+
+    Existing `DesktopAgentDetails` records MUST be used to limit discovery actions (in the next step) to the same mechanism as previously used or to skip the discovery step entirely if a `url` exists (immediately opening a hidden iframe).
+
+    If use of persisted data fails to establish a connection to the DA then `getAgent()` should reject its promise with "ReestablishConnectionFailed". 
+
+2) Create a race between these four mutually exclusive threads: 
+
+    **a) Timeout of 750ms** or the value provided by the `timeout` field
+
+    **b) Preload DAs:**
+
+    If a global (window or globalThis) `fdc3` object exists then return it immediately. If it does not exist then wait for the `fdc3Ready` event. After the `fdc3Ready` event triggers, recheck for the existence of the global `fdc3` object and return it if found.
+
+    **c) Browser-resident DAs:**
+
+    If a `url` exists from step (1) then proceed directly to step (iv), otherwise;
+
+    Locate all "Parents" (`window.opener`, `window.parent` or `window.parent.opener`). If no Parent exists then return null, otherwise simultaneously attempt to contact to each existing Parent with the following steps.
+
+    i) Add a listener (`<Parent>.addEventListener("message", (event) => {})`) to receive messages from each Parent.
+
+    ii) Send a "Handshake" message to each Parent (set for any origin since parent origin is unknown) with a unique nonce (one time random key).
+
+    iii) Accept the first correct response received (discard all others). Correct responses MUST include the original nonce. Stop the timer when a correct response is received. If the response has transferred a MessagePort (see HTML Channel Messaging API) then proceed to completion of step 2 (below). If the response contains a `url` then proceed to sub-step iv. If neither is provided then reject with "ErrorOnConnect".
+
+    iv) If a `url` has been provided, open a hidden iframe to the URL and listen on the `onload` event. Once the iframe is loaded, repeat steps i through iii substituting the hidden iframe for Parents.
+
+    **Race Resolution**
```

Comment:
I'm starting to wonder whether this should be in the spec as basically duplicative of the code we're going to ship in `@finos/fdc3`.  Maybe the spec should just say - you _must_ use that library?    By publishing the algorithm as well we are kind of inviting people to re-implement this logic (which we don't want) and also creating the problem that we have two sources of truth as to how it works, the code and this file.

Reply:
Technically, the code can't be the source of truth under our licensing - it has to be backed by a specification that forms the standard under the CSL, the code gets released under Apache 2.0. I think it also makes sense to explain whats going on under-the-hood.

We can't easily enforce the use of the original code easily (even if we refuse to document it here) as someone can always access it and reimplement or duplicate it. Neither the DA nor the FDC3 standard will have access to the internals of a sandboxed app in a web browser so theres just no way to know what code you are talking to over the postMessage/MessagePort. But with a spec we can at least test against that and require conformance to it. 

That said, I think it would be daft not to use the standard's own implementation of the proxy once published!